runtime/common/src/migrations/balances.rs

```diff
@@ -0,0 +1,269 @@
+	/// for transaction fee payment.
+	pub fee_frozen: Balance,
+}
+
+pub type OldAccountInfo<Index, Balance> = AccountInfo<Index, OldAccountData<Balance>>;
+
+pub type NewAccountData<Balance> = pallet_balances::AccountData<Balance>;
+
+pub type NewAccountInfo<Index, Balance> = AccountInfo<Index, NewAccountData<Balance>>;
+
+pub struct Migration<T: pallet_balances::Config + frame_system::Config>(
+	sp_std::marker::PhantomData<T>,
+);
+
+impl<T> OnRuntimeUpgrade for Migration<T>
+where
+	T: frame_system::Config<AccountId = AccountId32, Index = u32>
+		+ pallet_balances::Config<Balance = u128>,
+{
+	#[cfg(feature = "try-runtime")]
+	fn pre_upgrade() -> Result<Vec<u8>, DispatchError> {
+		let account_prefix = frame_system::Account::<T>::final_prefix();
+
+		let mut total_count = 0;
+		let mut total_reserved = 0;
+		let mut total_frozen = 0;
+
+		let mut previous_key = account_prefix.to_vec();
+
+		while let Some(next) = sp_io::storage::next_key(&previous_key) {
+			previous_key = next;
+
+			log::info!(
+				"Balances Migration - Processing account key - {}",
+				hex::encode(previous_key.clone())
+			);
+
+			// The difference between the old and new account data structures
+			// is the last field of the struct, which is:
+			//
+			// Old - `free_frozen` - T::Balance
+			// New - `flags` - u128
+			//
+			// During this check, we confirm that both the old and the new can
+			// be successfully decoded given the raw data found in storage, and
+			// we add specific checks for each version:
+			//
+			// Old - confirm that `fee_frozen` is zero, as it shouldn't have
+			// been used so far
+			// New - confirm that `flags` does not have the
+			// new logic flag set
+			match unhashed::get::<OldAccountInfo<T::Index, T::Balance>>(&previous_key) {
+				Some(old) => {
+					if !old.data.reserved.is_zero() {
+						total_reserved += 1;
+					}
+
+					if !old.data.fee_frozen.is_zero() || !old.data.misc_frozen.is_zero() {
+						total_frozen += 1;
+					}
+				}
+				None => log::error!("Balances Migration - Error decoding old data"),
+			};
+
+			match unhashed::get::<NewAccountInfo<T::Index, T::Balance>>(&previous_key) {
+				Some(new) => {
+					if new.data.flags.is_new_logic() {
+						log::warn!(
+							"Balances Migration - New account data with new logic flag enabled"
+						)
+					}
+				}
+				None => log::error!("Balances Migration - Error decoding new data"),
+			};
+
+			total_count += 1;
+		}
+
+		log::info!("Balances Migration - Total accounts - {}", total_count);
+		log::info!(
+			"Balances Migration - Total accounts with frozen balances - {}",
+			total_frozen
+		);
+		log::info!(
+			"Balances Migration - Total accounts with reserved balances - {}",
+			total_reserved
+		);
+
+		Ok(Vec::new())
+	}
+
+	fn on_runtime_upgrade() -> Weight {
+		// !!! REMOVE TEST ACCOUNT DATA !!!
+		//
+		// TEST ACCOUNT DATA - START
+		let test_account_data = get_test_account_data::<T>();
+
+		for (account_id, account_data) in test_account_data {
```

Comment:
@mustermeiszer @wischli do you think this should suffice or should we add more?

Reply:
I think that suffices. Thanks a lot!

Reply:
Note: `on_runtime_upgrade` already uses the new WASM sucht that we cannot write the old `AccountInfo`. Upgraded the test data to set new `AccountInfo` despite having a bit less purpose.

Reply:
I was just setting the storage with the raw encoded value of the old account data. How does the WASM version affect this?

Reply:
I don't have an answer for you but none of the accounts set during migration could be decoded whereas pre-existing ones with `reserved` or `misc_frozen` could. So it didn't simulate a real scenario.

Reply:
On that note: Putting the raw encoded data for the new AccountInfo also led to decoding issues. I don't have capacity to look into this. Writing via `pallet_balances::Account` worked and was sufficient to me given that this is just a testing migration.